rosbridge_library/src/rosbridge_library/internal/message_conversion.py

```diff
@@ -133,15 +137,23 @@ def extract_values(inst):
         raise InvalidMessageException(inst=inst)
     return _from_inst(inst, rostype)
 
+def decode_compressed_image(msg, inst):
+    inst.header.frame_id = msg["header"]["frame_id"]
+    inst.header.stamp.sec = msg["header"]["stamp"]["sec"]
+    inst.header.stamp.nanosec = msg["header"]["stamp"]["nanosec"]
+    inst.format = msg["format"]
+    inst.data = array.array("B", msg["data"])
+    return inst
 
 def populate_instance(msg, inst):
     """ Returns an instance of the provided class, with its fields populated
     according to the values in msg """
     inst_type = msg_instance_type_repr(inst)
+    if inst_type == "sensor_msgs/CompressedImage":
```

Comment:
I'm not convinced yet that hardcoding special handling for specific message types in this module in this way is a good idea.

If the data is coming in as anything but a base64 string, it shouldn't be coming anywhere near this code path.  I don't see any place where base64 decoding was added, so I don't understand how this decoder could possibly work.  Can you explain how this works?

Reply:
I agree that this is kind of a hacky solution to shortcut decoding of `CompressedImage`. At this point in the data flow, msg is a `dict` containing all the fields. In particular, arrays/lists have already been decoded into python/numpy arrays. From this point in the data flow on, the only remaining thing is to translate the fields from the dict into the attributes of `inst` (`inst` is already the target object of the correct type). in our debugging sessions it happened that decoding of CompressedImages was extremely slow. This is due to [an known problem](https://github.com/ros2/rosidl_python/issues/79) within the python-idl mapping. Particular, `inst.data = msg["data"]` would be the slow thing. Instead, it should be `inst.data = array.array("B", msg["data"])` which omits the runtime check if all values are within bounds. 

Since the generic decoding within `populate_instance` can only do `inst.data = msg["data"]`, I have included the shortcut here. 

Please let me know, if I should delete it in this fork or keep it.

Reply:
It seems like this could be made into a general solution by moving it into `_to_inst` with the rest of the decoding.  Otherwise I would accept this hack in this PR if there is a comment linking to [your comment explaining why it exists](https://github.com/RobotWebTools/rosbridge_suite/pull/493#discussion_r425151919).

Reply:
I have created an extra file that collects shortcut decoders. Currently, it only contains a decoder for `sensor_msgs/CompressedImage`. I needed that file to keep the `message_conversion.py` clean.  I have moved the call to these function into the `_to_object_inst`

Please see, if this solution is fine with you.